### ISSUE TITLE
Retrieving value fails, when using the `incr()` method with custom codec

### ISSUE THREAD

**Issue by pmeisen** [lettuce-io/lettuce-core#599]
When using a custom codec, e.g., the `SerializedObjectCodec` (as explained here https://github.com/lettuce-io/lettuce-core/wiki/Codecs) and utilizing the `incr()` method, the reading of the incremented value fails.

```
final RedisCommands<String, Object> sync = connection.sync();
sync.incr("counter");    // <---- this will only invoke the encodeKey() method of the codec
sync.get("counter");     // <---- fails in decodeValue(ByteBuffer bytes), because the bytes do not contain any serialized object, it contains the actual value of the counter
```

Any idea on how to solve this problem?

**Comment by mp911de** [comment]
You're describing an issue that is caused by encoding transformation. You're representing a Java object as a sequence of bytes that follows JDK serialization. Redis is binary-safe, meaning what you store is what you can retrieve (`GET`, `SET`). The Redis `INCR` command makes certain assumptions about your content: It expects a numeric value encoded using ASCII so `INCR` can convert the bytes to an integer value, increment it and encode it again with ASCII and store the bytes at the provided key. Redis transforms your bytes to a different encoding which is no longer readable by your codec.

Long story short: Use a `String`-based codec instead of the JDK serializer or create a `Number` codec if you rather want your API to exchange `Number` objects.

**Comment by pmeisen** [comment]
So if I have to store numbers (that can be incremented) and Java objects you recommend to use two connections, each utilizing a different codec (the number one could use the default codec, which I assume to work, for objects I use my custom codec)? A codec is connection based correct, so I have to utilize multiple connections, correct?

**Comment by mp911de** [comment]
That's the simple solution. 

A more sophisticated approach would use [custom commands](https://github.com/lettuce-io/lettuce-core/wiki/Custom-commands%2C-outputs-and-command-mechanics). It's a bit verbose but a good approach if you want to stick to a single connection and you can encapsulate the code behind some facade:

```java
AsyncCommand<String, String, String> set = new AsyncCommand<>(
        new Command<>(CommandType.SET,
                new ValueOutput<>(StringCodec.ASCII),
                new CommandArgs<>(StringCodec.ASCII).addKey("foo").addValue("1")));
redisConnection.dispatch((RedisCommand) set);

AsyncCommand<String, String, Long> incr = new AsyncCommand<>(
        new Command<>(CommandType.INCR,
                new IntegerOutput<>(StringCodec.ASCII),
                new CommandArgs<>(StringCodec.ASCII).addKey("foo")));
redisConnection.dispatch((RedisCommand) incr);

AsyncCommand<String, String, String> get = new AsyncCommand<>(
        new Command<>(CommandType.GET,
                new ValueOutput<>(StringCodec.ASCII),
                new CommandArgs<>(StringCodec.ASCII).addKey("foo")));
redisConnection.dispatch((RedisCommand) get);

System.out.println(get.join());
```

**Comment by mp911de** [comment]
May I close this ticket or is there anything else I can assist you with?

**Comment by mp911de** [comment]
Closing due to lack of requested feedback. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.

**Comment by pmeisen** [comment]
👍 the solution work just perfectly! Thank you so much for the support!
